http4k-security/digest/src/main/kotlin/org/http4k/security/digest/DigestChallenge.kt

```diff
@@ -0,0 +1,45 @@
+package org.http4k.security.digest
+
+import org.http4k.appendIf
+import org.http4k.appendIfPresent
+import org.http4k.security.Nonce
+
+/**
+ * The Digest Authorization challenge to be returned to the user as a header
+ */
+data class DigestChallenge(
+    val realm: String,
+    val nonce: Nonce,
+    val algorithm: String?,
+    val qop: List<Qop>,
+    val opaque: String?
+) {
+    fun toHeaderValue(): String {
+        val qopValue = qop.joinToString(", ") { it.value }
+
+        return StringBuilder("Digest realm=\"$realm\", nonce=\"$nonce\"")
+            .appendIfPresent(algorithm, ", algorithm=$algorithm")
+            .appendIf({ qop.isNotEmpty() }, ", qop=\"$qopValue\"")
+            .appendIfPresent(opaque, ", opaque=\"$opaque\"")
+            .toString()
+    }
+
+    companion object {
+        fun parse(headerValue: String): DigestChallenge? {
+            val (prefix, parameters) = ParameterizedHeader.parse(headerValue)
+
+            return when {
```

Comment:
Isn't this more complex than short-circuiting with an early return?

Reply:
Potentially - overall the project style is quite functional - one of expressions over multiple returns. If it's too much then happy to revert.

As for the fake data - yep - totally happy with that. It's the coverage that am most interested in. 🙃

Reply:
I can't commit to this branch, can I?  Shall I embed some new tests as a comment?

Reply:
You can pull the changes and then open a PR to this branch.

Reply:
Sure, makes sense; improvements are in #630